### PR TITLE
[1857] Complete and refactor subject editing for master subjects

### PR DESCRIPTION
**Pull request by faissalMT**
### Context
Users should be able to modify their subjects rather than having to go through support to do so.

### Changes proposed in this pull request
Allows the user to set single subjects.

### CODE REVIEW

**Comment by faissalMT** [comment]
[Edit Course 'Subject' (MCFE)](https://trello.com/c/JO6NoJWs/1857-edit-course-subject-mcfe)

File: app/controllers/courses/subjects_controller.rb [hunk 1]
```diff
@@ -7,7 +7,9 @@ class SubjectsController < ApplicationController
     def update
       master_subject = params.dig(:course, :master_subject)
 
-      if @course.update(subjects: [Subject.new(id: master_subject)])
+      subjects = @course.meta[:edit_options][:subjects].find { |subject| subject[:id] == master_subject }
```
Comment (misaka) [inline app/controllers/courses/subjects_controller.rb]: Looks like this should be `subject` not `subjects`?

File: app/controllers/courses/subjects_controller.rb [hunk 2]
```diff
@@ -7,7 +7,9 @@ class SubjectsController < ApplicationController
...
+
+      if @course.update(subjects: [Subject.new(subjects.to_h)])
```
[the remark this answers is not in the document]
Comment (misaka) [reply app/controllers/courses/subjects_controller.rb]: Designs from: https://bat-design-history.herokuapp.com/publish-teacher-training/new-course-iteration-14-jan#pick-a-subject

File: spec/decorators/course_decorator_spec.rb [hunk 3]
```diff
@@ -4,6 +4,10 @@
   let(:current_recruitment_cycle) { build :recruitment_cycle }
   let(:next_recruitment_cycle) { build :recruitment_cycle, :next_cycle }
   let(:provider) { build(:provider, accredited_body?: false) }
+  let(:english) { build(:subject, subject_name: "English") }
+  let(:mathematics) { build(:subject, subject_name: "Mathematics") }
```
Comment (misaka) [inline spec/decorators/course_decorator_spec.rb]: Would it makes sense for these to be traits on the `subject` factory? Seems to me that it'd reduce verbosity and chance or error.

File: spec/decorators/course_decorator_spec.rb [hunk 4]
```diff
@@ -167,4 +168,21 @@
       end
     end
   end
+
+  context "selectable master subjects" do
```
Comment (misaka) [inline spec/decorators/course_decorator_spec.rb]: I think you're describing `#selectable_master_subjects` here.

File: spec/features/courses/subjects/edit_spec.rb [hunk 5]
```diff
@@ -30,29 +30,45 @@
     stub_api_v2_resource(course, include: "subjects,sites,provider.sites,accrediting_provider")
   end
 
-  scenario "can visit the subject edit page" do
-    course_details_page.load_with_course(course)
-    expect(course_details_page).to be_displayed
-    course_details_page.edit_subjects_link.click
-    expect(subjects_page).to be_displayed
+  context "with given subjects" do
```
Comment (misaka) [inline spec/features/courses/subjects/edit_spec.rb]: Not sure what the difference is between this `"with given subjects"` context, and the `"with a given set if subjects"` context below? And it seems the way `course` is built here is essentially identical  to the one on line 14, so is this unnecessary extra noise?

File: spec/support/helpers.rb [hunk 6]
```diff
@@ -58,15 +58,16 @@ def stub_api_v2_request(url_path, stub, method = :get, status = 200, token: nil,
 
   def stub_api_v2_resource(resource,
                            jsonapi_response: nil,
-                           include: nil)
+                           include: nil,
+                           method: :get)
```
Comment (misaka) [inline spec/support/helpers.rb]: :100:

File: app/controllers/courses/subjects_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (faissalMT) [reply app/controllers/courses/subjects_controller.rb]: I've discussed this with @dankmitchell and he said that phase one will be just having the ability to set one subject.

[the remark this answers is not in the document]
Comment (misaka) [reply app/controllers/courses/subjects_controller.rb]: 👍

**Pull request merged**